src/windows/version.rs

```diff
@@ -0,0 +1,183 @@
+use std::mem::zeroed;
+use std::mem::size_of;
+
+extern crate winapi;
+use self::winapi::{ NTSTATUS, DWORD, STATUS_SUCCESS };
+use self::winapi::winnt::{ OSVERSIONINFOEXW };
+use self::winapi::sysinfoapi::{ SYSTEM_INFO };
+
+/// Win32 Flag: VER_NT_WORKSTATION
+///  https://msdn.microsoft.com/en-us/library/windows/desktop/ms724833(v=vs.85).aspx
+const VER_NT_WORKSTATION: u8  = 0x0000001;
+/// Win32 Flag: VER_SUITE_WH_SERVER
+///  https://msdn.microsoft.com/en-us/library/windows/desktop/ms724833(v=vs.85).aspx
+const VER_SUITE_WH_SERVER: u16 = 0x00008000;
+/// Win32 Flag: SM_SERVERR2
+/// https://msdn.microsoft.com/en-us/library/windows/desktop/ms724385(v=vs.85).aspx
+const SM_SERVERR2: u8  = 89;
+/// Win32 Flag: PROCESSOR_ARCHITECTURE_AMD64
+/// https://msdn.microsoft.com/en-us/library/windows/desktop/ms724958(v=vs.85).aspx
+const PROCESSOR_ARCHITECTURE_AMD64: u16 = 9;
+
+#[link(name = "ntdll")]
+extern {
+    pub fn RtlGetVersion(lpVersionInformation: &mut OSVERSIONINFOEXW) -> NTSTATUS;
+}
+
+#[link(name = "user32")]
+extern {
+    pub fn GetSystemMetrics(nIndex: u8) -> u8;
+}
+
+#[link(name = "kernel32")]
+extern {
+    pub fn GetSystemInfo(lpSystemInfo: &mut SYSTEM_INFO);
+}
+
+
+/// Win32Version Structure
+/// Holds information about the current systems version data and edition
+pub struct Win32Version {
+    pub osvi   : Option<OSVERSIONINFOEXW>,
+    pub edition: Option<String>
+}
+
+/// Win32Version Structure Implementation
+/// A set of functions that will set the data in the Win32Version struct
+/// osvi() will call edition() each time the function is called because a new
+/// structure is returned. However, edition() will only call osvi() if the
+/// osvi data has not been previously set.
+impl Win32Version {
+    pub fn osvi() -> Self {
+        unsafe {
+            let mut info: OSVERSIONINFOEXW = { zeroed() };
+            info.dwOSVersionInfoSize       = size_of::<OSVERSIONINFOEXW>() as DWORD;
+
+            if RtlGetVersion(&mut info) == STATUS_SUCCESS {
+                Self { osvi: Some(info), ..Default::default() }.edition()
+            } else {
+                Self { osvi: None, ..Default::default() }.edition()
+            }
+        }
+    }
+
+    pub fn edition(self) -> Self {
+        let mut info = match self.osvi {
+            Some(_) => self,
+            None    => Self::osvi()
+        };
+
+        let osvi = info.osvi.unwrap();
```

Comment:
What if `RtlGetVersion` fails? Then `osvi()` function will return `Win32Version` with `None` in the `osvi` field, right? After that `Win32Version::osvi` is going to be called unsuccessfully again which gives panic.

Am I missing something? If not - it would be better to handle it somehow.